layers/+lang/c-c++/funcs.el

```diff
@@ -12,151 +12,335 @@
+    (`ycmd (spacemacs//c-c++-setup-ycmd))))
+
+(defun spacemacs//c-c++-setup-company ()
+  "Conditionally setup C/C++ company integration based on backend."
+  (pcase (spacemacs//c-c++-backend)
+    (`lsp-clangd (spacemacs//c-c++-setup-lsp-company))
+    (`lsp-ccls (spacemacs//c-c++-setup-lsp-company))
+    (`lsp-cquery (spacemacs//c-c++-setup-lsp-company))
+    (`rtags (spacemacs//c-c++-setup-rtags-company))
+    (`ycmd (spacemacs//c-c++-setup-ycmd-company))))
+
+(defun spacemacs//c-c++-setup-dap ()
+  "Conditionally setup C/C++ DAP integration based on backend."
+  ;; currently DAP is only available using LSP
+  (pcase (spacemacs//c-c++-backend)
+    (`lsp-clangd (spacemacs//c-c++-setup-lsp-dap))
+    (`lsp-ccls (spacemacs//c-c++-setup-lsp-dap))
+    (`lsp-cquery (spacemacs//c-c++-setup-lsp-dap))))
+
+(defun spacemacs//c-c++-setup-eldoc ()
+  "Conditionally setup C/C++ eldoc integration based on backend."
+  (pcase (spacemacs//c-c++-backend)
+    ;; lsp setup eldoc on its own
+    (`ycmd (spacemacs//c-c++-setup-ycmd-eldoc))))
+
+(defun spacemacs//c-c++-setup-flycheck ()
+  "Conditionally setup C/C++ flycheck integration based on backend."
+  (pcase (spacemacs//c-c++-backend)
+    (`lsp-clangd (spacemacs//c-c++-setup-lsp-flycheck))
+    (`lsp-ccls (spacemacs//c-c++-setup-lsp-flycheck))
+    (`lsp-cquery (spacemacs//c-c++-setup-lsp-flycheck))
+    (`rtags (spacemacs//c-c++-setup-rtags-flycheck))
+    (`ycmd (spacemacs//c-c++-setup-ycmd-flycheck))))
+
+(defun spacemacs//c-c++-setup-format ()
+  "Conditionally setup format based on backend."
+  (pcase (spacemacs//c-c++-backend)
+    (`lsp-clangd (spacemacs//c-c++-setup-clang-format))
+    (`lsp-ccls (spacemacs//c-c++-setup-clang-format))
+    (`lsp-cquery (spacemacs//c-c++-setup-clang-format))))
 
 
-;; cpp-auto-include
-
-(defalias 'spacemacs/c++-organize-includes 'cpp-auto-include)
+;; lsp
 
-(defun spacemacs//c++-organize-includes-on-save ()
-  "Organize the includes on save when `c++-enable-organize-includes-on-save'
-is non-nil."
-  (when c++-enable-organize-includes-on-save
-    (spacemacs/c++-organize-includes)))
+;; clang
 
-(defun spacemacs/c++-organize-includes-on-save ()
-  "Add before-save hook for c++-organize-includes."
-  (add-hook 'before-save-hook
-            #'spacemacs//c++-organize-includes-on-save nil t))
+(defun spacemacs//c-c++-setup-lsp-clangd ()
+  "Setup LSP clangd."
+  ;; extensions
+  (spacemacs/lsp-define-extensions
+   "c-c++" "lsp-clangd"
+   'clangd-other-file "textDocument/switchSourceHeader" 'buffer-file-name)
+  (set (make-local-variable 'lsp-enabled-clients) '(clangd))
+  (lsp))
+
+;; ccls
+
+(defun spacemacs//c-c++-setup-lsp-ccls ()
+  "Setup LSP ccls."
+  (require 'ccls)
+  ;; configuration
+  (let ((cache-dir
+         (if (null c-c++-lsp-cache-dir)
+             (concat spacemacs-cache-directory "lsp-ccls")
+           (concat (file-truename (file-name-as-directory c-c++-lsp-cache-dir))
+                   "lsp-ccls"))))
```

Comment:
ccls doesn't require `cache-dir` to be defined as it uses `.ccls-cache` default directory, similar to `.clangd`. With this scheme it is impossible to set it to a default value by not defining it. I'd frankly let a user set it through `ccls-initialization-options` instead as only cquery requires this setting and clangd does not even allow to change it.

Reply:
I made `c-c++-lsp-cache-directory` a `cquery` only variable by renaming it `c-c++-lsp-cquery-cache-directory`. I updated the README.org file to add the information you mentioned.
Tell me if this sounds good to you now.

Reply:
Looks good!